File: python/Q1046_last_stone_weight.py
#
# @lc app=leetcode id=1046 lang=python3
#
# [1046] Last Stone Weight
#


# @lc code=start
import heapq


class Solution:
    def lastStoneWeight(self, stones: list[int]) -> int:
        stone_list: list[int] = [-s for s in stones]

        heapq.heapify(stone_list)

        while len(stone_list) > 1:
            larger: int = heapq.heappop(stone_list)
            smaller: int = heapq.heappop(stone_list)

            if larger != smaller:
                heapq.heappush(stone_list, larger - smaller)

        return -stone_list[0] if stone_list else 0


# @lc code=end

File: rust/src/leetcode/Q1046_last_stone_weight.rs
/*
 * @lc app=leetcode id=1046 lang=rust
 *
 * [1046] Last Stone Weight
 */

// @lc code=start
use std::collections::BinaryHeap;

impl Solution {
    pub fn last_stone_weight(stones: Vec<i32>) -> i32 {
        let mut heap: BinaryHeap<i32> = BinaryHeap::from(stones);

        while heap.len() > 1 {
            let stone1: i32 = heap.pop().unwrap();
            let stone2: i32 = heap.pop().unwrap();

            if stone1 != stone2 {
                heap.push(stone1 - stone2);
            }
        }

        if !heap.is_empty() {
            *heap.peek().unwrap()
        } else {
            0
        }
    }
}
// @lc code=end
